### ISSUE TITLE
Implement PPO Neural Policy, Critic Network, and Full PPO Update Logic

### ISSUE THREAD

**Issue by athariandre** [athariandre/yasta#5]
Goal

Replace the placeholder DummyPolicy with a fully functional PyTorch Actor–Critic network, integrate it into the PR-2 training pipeline, implement the complete PPO update step, add model checkpoint saving/loading, and validate the entire training loop end-to-end.

This PR must not modify:

agent.py

case_closed_game.py

judge_engine.py

Only training/ files may be changed or added.

✅ Files to Create / Modify
✅ 1. training/policy.py (new file)

Implement a CPU-only PyTorch module containing:

A. Shared encoder

Input:

board: (H, W) float32

my_head, opp_head: (2,)

boosts: (2,)

turns: (1,)

directions: (2,)

alive: (2,)

Output: latent feature vector of size 128.

B. Actor head

Input: latent vector

Output: categorical logits over 4 actions

Must use:

small MLP (≤2 layers)

no convolutions (to keep CPU inference fast)

C. Critic head

Input: latent vector

Output: scalar value estimate

D. Methods

act(obs)

returns (action, logprob, value)

evaluate_actions(observations, actions)

returns (logprobs, values, entropy)

E. CPU-only training safety

Always call .to(torch.device('cpu'))

Never reference CUDA

F. Model size constraint

Must ensure <5MB total parameters (safe for Docker)

Prefer hidden sizes in [128, 256]

✅ 2. Modify training/trainer.py

Replace _ppo_update() with full PPO logic:

A. Mini-batch PPO loop

For each epoch:

Shuffle indices

Fetch mini-batches

Call policy.evaluate_actions()

Compute:

policy loss (clip objective)

value loss (MSE)

entropy loss

Combine using config:

PPO_CLIP_EPS

VALUE_COEFF

ENTROPY_COEFF

Backpropagate

Clip gradients to config.MAX_GRAD_NORM

Optimizer step

B. Add optimizer initialization in __init__

Example components:

torch.optim.Adam

learning rate from config

C. Save & load checkpoints

In _save_checkpoint():

Add model state dict

Add optimizer state dict

In load_checkpoint():

Restore both

D. Determinism

Call:

torch.manual_seed

ensure no CuDNN usage (CPU-only)

✅ 3. Modify training/rollout.py

Small updates only:

✅ Add ability to flatten observation dicts into tensor batches for policy evaluation.
✅ Ensure all observations remain CPU numpy → CPU torch.

✅ 4. Modify training/env.py

No logic changes, only:

✅ Convert all obs arrays to np.float32
✅ Guarantee consistent shapes
✅ Add inline comment: “PR-3: observations must be stable for NN input”

✅ 5. Modify training/config.py

Add PPO hyperparameters:

# network sizes
LATENT_SIZE = 128

# PPO
PPO_CLIP_EPS = 0.2
NUM_EPOCHS = 4
BATCH_SIZE = 512
VALUE_COEFF = 0.5
ENTROPY_COEFF = 0.01
MAX_GRAD_NORM = 0.5

# learning rate
LEARNING_RATE = 3e-4

Include:

DEVICE = "cpu"

OBS_NORMALIZATION_EPS = 1e-8

✅ 6. Modify training/readme.md

Add documentation sections:

✅ Neural network architecture
✅ PPO update explanation
✅ How checkpoint saving/loading works
✅ Expected training stability behavior
✅ CPU-only constraints reminder

✅ Behavioral Requirements
✅ Policy Behavior

Your LLM agent must implement:

safe feature extraction

stable PPO optimization

correct advantage normalization

consistent shaping rewards

✅ Training Stability

The model must never collapse due to:

exploding gradients

NaNs

invalid log probabilities

unbounded value loss

✅ Inference Speed Requirement

Model inference must stay:

under 5ms per step

on CPU only

using only small MLP layers

This ensures the trained model can run inside the competition’s tight tick limit.

✅ Testing Requirements
1. training/validate_pr3.py (new file)

Validate:

imports succeed

policy builds successfully

forward pass works

PPO update runs for 1 fake batch

checkpoint save/load works

This script must use very small synthetic batches to avoid slowdown.

2. Add a CI-style printout in trainer:

After each PPO update print:

[Update X] Policy loss: …
              Value loss: …
              Entropy: …
              Clip frac: …

✅ Non-Negotiable Constraints

No GPU, no CUDA, no cuDNN.

No TensorFlow, JAX, or large ML frameworks.

Model must remain <5MB serialized.

Inference must be fast enough for real matches.

Do NOT change the competition environment logic.

✅ Deliverables Summary
New Files

training/policy.py

training/validate_pr3.py

Modified Files

training/trainer.py

training/rollout.py

training/env.py (small adjustments)

training/config.py

training/readme.md

No Touch Files

agent.py

case_closed_game.py

judge_engine.py
